philips-labs/repolinter
# Add option to create Github issue with 'fixers'

## Description

We want to be able to 'fix' a broken rule and be able to configure it per rule what should happen. For some rules, a notification should be sent, others a Github Issue should be created, some should even create a PR.
This Pull Request adds the ability to define a fix that will create a customizable issue on a target repository.
  
This currently includes:
- Each rule includes a unique id that allows us to programmatically identify which issue is part of which rule.
- Checks if there are existing issues that are open to prevent issue spam. If there are open issues, matching the unique rule id, it will do nothing.
- Checks if there are existing closed issues that match the unique rule id. If there are, it will re-open them, update the issue body/title and comment on it. This makes the repo owner aware that there is some sort of regression.
- Automatically create and assign labels like: 'Continuous Compliance', 'automated', 'CC: Bypass'
- Add an ability to bypass future CC checks for a specific ruleset by applying the Bypass Label.
- Each rule configuration is customizable within the ruleset (https://github.com/Brend-Smits/octokit-test-repo/blob/main/ruleset.json#L37-L46)
- Environment variables that are available and required: TARGET_REPO (philips-labs/repolinter) GITHUB_TOKEN (Personal Access Token)

## To-do

- [x] The Personal Access Token(PAT) is currently hardcoded in the file, retrieve it from environment variables.
- [x] Write tests to verify functionality.
- [x] Allow dynamically setting of target repository  where the issue will be created (https://github.com/philips-labs/repolinter/blob/feature/github-issue-fix/fixes/github-issue-create.js#L23-L24) **Help is appreciated here 😁**  
I've temporarily resolved this by allowing you to set a TARGET_REPO environment variable. This is not ideal, but I was unable to retrieve the Git URL from somewhere else. 
- [x] Return proper exit codes, when a fix is applied the application did not fail.

## Test Plan

1. Input a PAT via environment variable using ``export GITHUB_TOKEN=TOKENHERE``
2. Input a target repository via environment variable using ``export TARGET_REPO=brend-smits/octo-test-repo``
2. Execute ``./bin/repolinter.js -g https://github.com/Brend-Smits/octokit-test-repo --rulesetFile ruleset.json`` or execute it on your own repository with your own ruleset.
2. Execute the repolinter with a rule you know will fail/succeed, check if a Github issue is created on the above test repository. (Why on the test repository? Please refer to the above To-do section.
3. Close the newly created issue and execute repolinter again, it should re-open the old issue if the same rule was triggered.
4. Apply the Bypass Label to an issue and verify that the repolinter will not create a new issue for the same rule.


Feedback is much appreciated, I'm not very proficient with Javascript and honestly not amused with the lack of types, classes, and interfaces 🙄
